Use std::span more in AtomString / AtomStringImpl

cdumez opened the pull request:
#### fca7a271ab7afb009bfc5f324cc0a9a5a333d160
<pre>
Use std::span more in AtomString / AtomStringImpl
<a href="https://bugs.webkit.org/show_bug.cgi?id=272174">https://bugs.webkit.org/show_bug.cgi?id=272174</a>

Reviewed by Brent Fulgham and Ryosuke Niwa.

* Source/WTF/wtf/text/AtomString.cpp:
(WTF::AtomString::fromUTF8Internal):
* Source/WTF/wtf/text/AtomString.h:
(WTF::AtomString::fromUTF8):
* Source/WTF/wtf/text/AtomStringImpl.cpp:
(WTF::AtomStringImpl::addLiteral):
(WTF::AtomStringImpl::addUTF8):
* Source/WTF/wtf/text/AtomStringImpl.h:
* Source/WTF/wtf/text/icu/TextBreakIteratorICU.h:
(WTF::TextBreakIteratorICU::makeLocaleWithBreakKeyword):
* Source/WTF/wtf/unicode/UTF8Conversion.cpp:
(WTF::Unicode::calculateStringHashAndLengthFromUTF8MaskingTop8Bits):
* Source/WTF/wtf/unicode/UTF8Conversion.h:
* Source/WebCore/html/parser/HTMLMetaCharsetParser.cpp:
(WebCore::HTMLMetaCharsetParser::checkForMetaCharset):
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::processAttributes):
(WebCore::HTMLPreloadScanner::scan):
* Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.cpp:
(WebCore::AudioTrackPrivateWebM::label const):
(WebCore::AudioTrackPrivateWebM::language const):
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.cpp:
(WebCore::VideoTrackPrivateWebM::label const):
(WebCore::VideoTrackPrivateWebM::language const):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::toAtomString):

Canonical link: <a href="https://commits.webkit.org/277087@main">https://commits.webkit.org/277087@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14a4c448da77941f847176cf608a6030fdfa0776

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46664 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25824 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49274 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49341 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42715 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48971 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30184 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23286 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47244 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/22803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19317 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/20181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/41332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4709 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/39911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42948 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/41696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/51216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/46146 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21671 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/51216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/51216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/53289 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6525 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22664 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/10935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WTF/wtf/text/AtomString.cpp:
```diff
@@ -117,15 +117,10 @@ AtomString AtomString::number(double number)
-    ASSERT(end || start[0]);
-
-    return AtomStringImpl::addUTF8(start, end ? end : start + std::strlen(start));
+    ASSERT(!characters.empty());
```
Should this be a RELEASE_ASSERT?

cdumez replied:
It wasn't previously and I don't think we'd have a security bug in release if the assertion didn't hold. As a result, I'm included to keep as debug to avoid hurting perf in a refactoring patch.

brentfulgham commented on Source/WTF/wtf/text/AtomStringImpl.cpp:
```diff
@@ -329,12 +329,11 @@ RefPtr<AtomStringImpl> AtomStringImpl::add(std::span<const LChar> characters)
 {
-    ASSERT(characters);
-    ASSERT(length);
+    ASSERT(!characters.empty());
```
To be pedantic: Shouldn't we be using `ASSERT(characters.data())`, too? The original asserts looked for zero-length strings, and strings with a nullptr first character.

cdumez replied:
I can add `ASSERT(characters.data())` if you'd like. I personally didn't think it was critical to have.

cdumez committed:
Use std::span more in AtomString / AtomStringImpl
https://bugs.webkit.org/show_bug.cgi?id=272174

Reviewed by Brent Fulgham and Ryosuke Niwa.

* Source/WTF/wtf/text/AtomString.cpp:
(WTF::AtomString::fromUTF8Internal):
* Source/WTF/wtf/text/AtomString.h:
(WTF::AtomString::fromUTF8):
* Source/WTF/wtf/text/AtomStringImpl.cpp:
(WTF::AtomStringImpl::addLiteral):
(WTF::AtomStringImpl::addUTF8):
* Source/WTF/wtf/text/AtomStringImpl.h:
* Source/WTF/wtf/text/icu/TextBreakIteratorICU.h:
(WTF::TextBreakIteratorICU::makeLocaleWithBreakKeyword):
* Source/WTF/wtf/unicode/UTF8Conversion.cpp:
(WTF::Unicode::calculateStringHashAndLengthFromUTF8MaskingTop8Bits):
* Source/WTF/wtf/unicode/UTF8Conversion.h:
* Source/WebCore/html/parser/HTMLMetaCharsetParser.cpp:
(WebCore::HTMLMetaCharsetParser::checkForMetaCharset):
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::processAttributes):
(WebCore::HTMLPreloadScanner::scan):
* Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.cpp:
(WebCore::AudioTrackPrivateWebM::label const):
(WebCore::AudioTrackPrivateWebM::language const):
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.cpp:
(WebCore::VideoTrackPrivateWebM::label const):
(WebCore::VideoTrackPrivateWebM::language const):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::toAtomString):

Canonical link: https://commits.webkit.org/277087@main
